Add rekognition-service

Manuelacosta98 opened the pull request:
Hi great library, i am trying to add rekognition service to async-aws, i encounter a problem with the method  SearchFacesByImage  the problem is the next:

`In RestJsonSerializer.php line 158:
                                     
  Type float is not yet implemented `

I am unable to track the issue, i am a newbie in OSS but want to get this done, Thanks for the help i will start adding the test in the meanwhile

Manuelacosta98 commented:
> Hi great library, i am trying to add rekognition service to async-aws, i encounter a problem with the method SearchFacesByImage the problem is the next:
> 
> `In RestJsonSerializer.php line 158:
> 
> Type float is not yet implemented `
> 
> I am unable to track the issue, i am a newbie in OSS but want to get this done, Thanks for the help i will start adding the test in the meanwhile

The problem seams to be in the function dumpArrayElement in RestJsonSerializer

` 
        switch ($shape->getType()) {
            case 'string':
            case 'integer':
            case 'float':
            case 'long':
                return $this->dumpArrayScalar($output, $input, $contextProperty, $shape);
            case 'boolean':
                return $this->dumpArrayBoolean($output, $input, $shape);
            case 'timestamp':
                return $this->dumpArrayTimestamp($output, $input, $shape);
            case 'blob':
                return $this->dumpArrayBlob($output, $input, $shape, $isRequired);
        }`

I add the float type to the scalar value case in the function.

So if someone can take a look of this i will be testing it soon.

jderusse commented:
Thanks for your contribution. Looks like your fix is OK. We did not implemented `FLOAT` in Json because we didn't had use case for it.
Could you confirm that passing a float in `FaceMatchThreshold` to `SearchFacesByImage` is accepted by AWS?
and calling `getSearchedFaceConfidence` is a valid float in the repsonse?

You'll need to rebase your PR (fetch last version) and run `generate --all` to fix `sts`/`sqs` coding standard
And implements tests that have been partially generated.

Manuelacosta98 commented:
> Thanks for your contribution. Looks like your fix is OK. We did not implemented `FLOAT` in Json because we didn't had use case for it.
> Could you confirm that passing a float in `FaceMatchThreshold` to `SearchFacesByImage` is accepted by AWS?
> and calling `getSearchedFaceConfidence` is a valid float in the repsonse?
> 
> You'll need to rebase your PR (fetch last version) and run `generate --all` to fix `sts`/`sqs` coding standard
> And implements tests that have been partially generated.

Hi @jderusse thanks for replying i will give it a try today and answer. i dint get the rebase part you got a tag or branch on witch i need to rebase? 

Sorry i fork the repo from master i am new to these.

jderusse commented:
see the warning "This branch has conflicts that must be resolved"  in the GitHub interface.
it means that somebody else pushed in master (another PR) and your fork is not up to date.
Given the conflicted files are not related I think running `generate --all` should fix the issue and you don't have to rebase.

but if you are curious and want more information about rebase and fork, have a look to https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

Manuelacosta98 commented:
> see the warning "This branch has conflicts that must be resolved" in the GitHub interface.
> it means that somebody else pushed in master (another PR) and your fork is not up to date.
> Given the conflicted files are not related I think running `generate --all` should fix the issue and you don't have to rebase.
> 
> but if you are curious and want more information about rebase and fork, have a look to https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

Hi @jderusse sorry for the time past i got two problems the first is a question who do i run the test only in the new service? , and the other is that i try to require my fork repo in a project to test it and i encounter with this error 
![Screenshot from 2020-08-13 10-46-18](https://user-images.githubusercontent.com/33484814/90157066-e0651880-dd52-11ea-9386-951f3e1a829f.png)

is there a way to include via composer the recognition client from my fork repo? 

Thanks a lot

Manuelacosta98 commented:
> > see the warning "This branch has conflicts that must be resolved" in the GitHub interface.
> > it means that somebody else pushed in master (another PR) and your fork is not up to date.
> > Given the conflicted files are not related I think running `generate --all` should fix the issue and you don't have to rebase.
> > but if you are curious and want more information about rebase and fork, have a look to https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request
> 
> Hi @jderusse sorry for the time past i got two problems the first is a question who do i run the test only in the new service? , and the other is that i try to require my fork repo in a project to test it and i encounter with this error
> ![Screenshot from 2020-08-13 10-46-18](https://user-images.githubusercontent.com/33484814/90157066-e0651880-dd52-11ea-9386-951f3e1a829f.png)
> 
> is there a way to include via composer the recognition client from my fork repo?
> 
> Thanks a lot

When i try to run the individual test with my global phpunit i get these error:

 "PHP Fatal error:  Uncaught Error: Class 'AsyncAws\Core\Test\TestCase' not found"

I already run the composer install , theres another step to take before testing or only can be done with the docker makefile?

Manuelacosta98 commented:
Ok i found out how is done, i am fixing the falling test, Thanks

Manuelacosta98 commented:
Just keeping the updates i implemented the request unit test in the new service an also added some basic documentation to pass those test as well, tomorrow i will implement the test in the unit response side, i do not have knowledge in docker container i think some of these test can only be implemented that way i am not quite sure.

Thanks a lot for the follow up.

jderusse commented on src/Service/Rekognition/.github/workflows/ci.yml:
```diff
@@ -0,0 +1,40 @@
+    strategy:
+      max-parallel: 10
+      matrix:
+        php: ['7.2', '7.3', '7.4']
```
missing php8? 

/cc @Nyholm

Nyholm commented on composer.json:
```diff
@@ -67,7 +67,8 @@
             "AsyncAws\\SimpleS3\\": "src/Integration/Aws/SimpleS3/src",
-            "AsyncAws\\Symfony\\Bundle\\": "src/Integration/Symfony/Bundle/src"
+            "AsyncAws\\Symfony\\Bundle\\": "src/Integration/Symfony/Bundle/src",
+            "AsyncAws\\Rekognition\\": "src/Service/Rekognition/src"
```
Put it just over S3 to keep this in alphabetical order and next to the other clients.

Nyholm commented on composer.json:
```diff
@@ -96,7 +97,8 @@
             "AsyncAws\\SimpleS3\\Tests\\": "src/Integration/Aws/SimpleS3/tests",
-            "AsyncAws\\Symfony\\Bundle\\Tests\\": "src/Integration/Symfony/Bundle/tests"
+            "AsyncAws\\Symfony\\Bundle\\Tests\\": "src/Integration/Symfony/Bundle/tests",
+            "AsyncAws\\Rekognition\\Tests\\": "src/Service/Rekognition/tests"
```
Same here

Nyholm commented on couscous.yml:
```diff
@@ -67,6 +67,10 @@ menu:
             sts:
                 text: STS
                 url: /clients/sts.html
+            rekognition:
```
Same here, Just above S3

Nyholm commented on docs/clients/index.md:
```diff
@@ -149,3 +149,4 @@ for more information.
 | [SQS](./sqs.md)                             | [async-aws/sqs](https://packagist.org/packages/async-aws/sqs)
 | [SSM](./ssm.md)                             | [async-aws/ssm](https://packagist.org/packages/async-aws/ssm)
 | [STS](./sts.md)                             | [async-aws/core](https://packagist.org/packages/async-aws/core)
+| [Rekognition](./rekognition.md)             | [async-aws/rekognition](https://packagist.org/packages/async-aws/rekognition)
```
Same

Nyholm commented on docs/clients/rekognition.md:
```diff
@@ -0,0 +1,77 @@
+
+$response = $rekognition->createCollection($request);
+
+$response->resolve();
```
This line is not needed.

Nyholm commented on docs/clients/rekognition.md:
```diff
@@ -0,0 +1,77 @@
+
+$rekognition = new RekognitionClient();
+
+$addFacesRequest = new searchFacesByImageRequest(
```
```suggestion
$addFacesRequest = new SearchFacesByImageRequest(
```

Nyholm commented on src/Core/src/AwsClientFactory.php:
```diff
@@ -258,4 +259,17 @@ public function cognitoIdentityProvider(): CognitoIdentityProviderClient
         return $this->serviceCache[__METHOD__];
     }
+
+    public function rekognition(): RekognitionClient
```
Could you put this just over S3?

Nyholm commented on src/Integration/Symfony/Bundle/src/DependencyInjection/AwsPackagesProvider.php:
```diff
@@ -69,6 +69,10 @@ public static function getAllServices(): array
                 'class' => \AsyncAws\Core\Sts\StsClient::class,
                 'package' => 'async-aws/core',
             ],
+            'rekognition' => [
```
Same

Nyholm commented on src/Integration/Symfony/Bundle/src/DependencyInjection/AwsPackagesProvider.php:
```diff
@@ -69,6 +69,10 @@ public static function getAllServices(): array
             ],
+            'rekognition' => [
+                'class' => \AsyncAws\Rekognition\RekognitionClient::class,
+                'package' => 'async-aws/Rekognition',
```
```suggestion
                'package' => 'async-aws/rekognition',
```

Nyholm commented on src/Service/Rekognition/tests/Unit/Result/DeleteProjectResponseTest.php:
```diff
@@ -0,0 +1,30 @@
+{
+    public function testDeleteProjectResponse(): void
+    {
+        self::fail('Not implemented');
```
Make sure we complete all these tests

Nyholm commented on README.md:
```diff
@@ -17,13 +17,14 @@ See full documentation on [https://async-aws.com](https://async-aws.com).
 | [async-aws/dynamo-db](https://github.com/async-aws/dynamo-db)                                 | [![Latest Stable Version](https://poser.pugx.org/async-aws/dynamo-db/v/stable)](https://packagist.org/packages/async-aws/dynamo-db)                                 [![Total Downloads](https://poser.pugx.org/async-aws/dynamo-db/downloads)](https://packagist.org/packages/async-aws/dynamo-db)            ...
 | [async-aws/event-bridge](https://github.com/async-aws/event-bridge)                           | [![Latest Stable Version](https://poser.pugx.org/async-aws/event-bridge/v/stable)](https://packagist.org/packages/async-aws/event-bridge)                           [![Total Downloads](https://poser.pugx.org/async-aws/event-bridge/downloads)](https://packagist.org/packages/async-aws/event-bridge)      ...
-| [async-aws/iam](https://github.com/async-aws/iam)                                             | [![Latest Stable Version](https://poser.pugx.org/async-aws/iam/v/stable)](https://packagist.org/packages/async-aws/iam)                                             [![Total Downloads](https://poser.pugx.org/async-aws/iam/downloads)](https://packagist.org/packages/async-aws/iam)                        ...
+| [async-aws/iam](https://github.com/async-aws/iam)                                             | [![Latest Stable Version](https://poser.pugx.org/async-aws/iam/v/stable)](https://packagist.org/packages/async-aws/iam)                                             [![Total Downloads](https://poser.pugx.org/async-aws/iam/downloads)](https://packagist.org/packages/async-aws/iam)                        ...
```
We should not change this line.

I dont really see what's changed to be honest.

Nyholm commented on src/Service/Rekognition/.github/workflows/branch_alias.yml:
```diff
@@ -0,0 +1,76 @@
+
+      - name: Update branch alias
+        run: |
+          cd aws/src/Service/Foobar
```
```suggestion
          cd aws/src/Service/Rekognition
```

Nyholm replied:
This is why the rootTest fails.

Nyholm reviewed (approved):
Thank you

Nyholm commented:
The new psalm failures are false positive. 

Psalm says: "We cannot trust the data coming form the server". Which we really can since we have a contract.

jderusse commented:
thank you @Nyholm for your help.

Nyholm commented:
Thank you @jderusse and @Manuelacosta98 

https://packagist.org/packages/async-aws/rekognition

Manuelacosta98 commented:
Thanks for all i got pretty sick sorry about that

Nyholm commented:
We are happy to help. 

I hope you have recovered and feel better now. Your health is always more important than some open source work.
